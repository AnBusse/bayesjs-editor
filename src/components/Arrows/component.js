import React from 'react';

import PropTypes from 'prop-types';
import { arrowsPropTypes } from 'models';
import Arrow from '../Arrow';
import ArrowsDefs from '../ArrowsDefs';

const Arrows = ({
  arrows,
  keyFocus,
  onMouseOver,
  onMouseLeave,
  getStrokeOpacity,
  getMarkEndStyle,
  onMouseDown,
}) => (
  <g>
    <ArrowsDefs />
    {arrows.map(({
      key, from, to, markEnd, childId, parentId,
    }) => ( // eslint-disable-next-line jsx-a11y/mouse-events-have-key-events
      <Arrow
        key={key}
        id={key}
        onMouseOver={onMouseOver}
        onMouseLeave={onMouseLeave}
        strokeOpacity={getStrokeOpacity(key, keyFocus)}
        markEndStyle={getMarkEndStyle(key, keyFocus)}
        from={from}
        to={to}
        markEnd={markEnd}
        onMouseDown={onMouseDown({ childId, parentId })}
      />
    ))}
  </g>
);

Arrows.propTypes = {
  arrows: arrowsPropTypes.isRequired,
  keyFocus: PropTypes.string.isRequired,
  onMouseOver: PropTypes.func.isRequired,
  onMouseLeave: PropTypes.func.isRequired,
  getStrokeOpacity: PropTypes.func.isRequired,
  getMarkEndStyle: PropTypes.func.isRequired,
  onMouseDown: PropTypes.func.isRequired,
};

export default Arrows;
