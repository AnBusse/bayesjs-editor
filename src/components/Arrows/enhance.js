import {
  compose, pure, withStateHandlers, withHandlers,
} from 'recompose';
import { equals, isNil } from 'ramda';
import { isFunction } from 'lodash';

const isKeyFocused = (key, keyFocus) => isNil(keyFocus) || equals(key, keyFocus);

const enhance = compose(
  pure,
  withStateHandlers(
    ({ initialKeyFocus = null }) => ({
      keyFocus: initialKeyFocus,
    }),
    {
      onMouseOver: () => keyFocus => ({
        keyFocus,
      }),
      onMouseLeave: (_, { initialKeyFocus }) => () => ({
        keyFocus: initialKeyFocus,
      }),
    },
  ),
  withHandlers({
    getStrokeOpacity: () => (key, keyFocused) => isKeyFocused(key, keyFocused) ? 1 : 0.2,
    getMarkEndStyle: () => (key, keyFocused) =>
      isKeyFocused(key, keyFocused)
        ? 'url(#triangle)'
        : 'url(##triangle-with-low-opacity)',
    onMouseDown: ({ onMouseDown }) => opts => e =>
      isFunction(onMouseDown)
        ? onMouseDown(e, opts)
        : null,
  }),
);

export default enhance;
