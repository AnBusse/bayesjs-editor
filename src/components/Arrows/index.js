import React, { Component } from 'react';

import PropTypes from 'prop-types';

class Arrows extends Component {
  constructor(props) {
    super(props);

    this.state = {
      indexFocus: null,
    };
  }

  renderDefs = () => (
    <defs>
      <marker
        id="triangle"
        viewBox="0 0 10 10"
        markerWidth="6"
        markerHeight="6"
        refX="8"
        refY="5"
        orient="auto"
      >
        <path d="M0,0 L10,5 L0,10" fill="#333" />
      </marker>

      <marker
        id="triangle-opacity"
        viewBox="0 0 10 10"
        markerWidth="6"
        markerHeight="6"
        refX="8"
        refY="5"
        orient="auto"
        opacity="0.2"
      >
        <path d="M0,0 L10,5 L0,10" fill="#333" />
      </marker>
    </defs>
  );

  bringArrowToFront = uuid => () => {
    const { arrowsRendered } = this.state;
    const sorted = arrowsRendered.sort((a, b) => {
      if (a.key === uuid) return 1;
      if (b.key === uuid) return -1;
      return 0;
    });

    this.setState({
      arrowsRendered: sorted,
    });
  };

  onMouseOver = index => () => {
    this.setState({
      indexFocus: index,
    });
  };

  onMouseLeave = () => () => {
    this.setState({
      indexFocus: null,
    });
  };

  getStrokeOpacity = (index, indexFocus) => {
    if (indexFocus != null) {
      if (index !== indexFocus) return 0.2;
    }

    return 1;
  };

  getArrow = (arrow, index, indexFocus) => {
    const value = this.getStrokeOpacity(index, indexFocus);

    return React.cloneElement(
      arrow,
      { markEndStyle: value === 1 ? 'url(#triangle)' : 'url(#triangle-opacity)' },
    );
  }

  render() {
    const { arrows } = this.props;
    const { indexFocus } = this.state;

    return (
      <g>
        {this.renderDefs()}
        {arrows.map((arrow, i) => (
          <g
            key={i} // eslint-disable-line
            onMouseOver={this.onMouseOver(i)}
            onFocus={this.onMouseOver(i)}
            onMouseLeave={this.onMouseLeave(i)}
            strokeOpacity={this.getStrokeOpacity(i, indexFocus)}
            style={{
              transition: 'stroke-opacity 0.2s',
            }}
          >
            {this.getArrow(arrow, i, indexFocus)}
          </g>
        ))}
      </g>
    );
  }
}

Arrows.propTypes = {
  arrows: PropTypes.arrayOf(PropTypes.element).isRequired,
};

export default Arrows;
