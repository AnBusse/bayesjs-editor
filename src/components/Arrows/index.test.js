import React from 'react';
import { shallow } from 'enzyme';
import Arrows from './component';

jest.mock('components/Arrow', () => {
  const Arrow = () => <div />;
  return Arrow;
});

const defaultProps = {
  arrows: [{
    key: 'arrow-1',
  }, {
    key: 'arrow-2',
  }, {
    key: 'arrow-3',
  }],
  getStrokeOpacity: () => 1,
  getMarkEndStyle: () => 'url(#triangle)',
  onMouseDown: jest.fn(),
  onMouseLeave: jest.fn(),
  keyFocus: '',
};

const shallowComponent = (props = {}) => {
  const compProps = { ...defaultProps, ...props };

  return shallow(<Arrows {...compProps} />);
};

describe('Arrows Component', () => {
  it('matches snapshot', () => {
    const component = shallowComponent();

    expect(component).toMatchSnapshot();
  });
});
