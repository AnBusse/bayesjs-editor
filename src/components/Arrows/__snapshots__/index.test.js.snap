// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Arrows Component matches snapshot 1`] = `
<g>
  <ArrowsDefs />
  <Arrow
    id="arrow-1"
    key="arrow-1"
    markEndStyle="url(#triangle)"
    onMouseLeave={[Function]}
    strokeOpacity={1}
  />
  <Arrow
    id="arrow-2"
    key="arrow-2"
    markEndStyle="url(#triangle)"
    onMouseLeave={[Function]}
    strokeOpacity={1}
  />
  <Arrow
    id="arrow-3"
    key="arrow-3"
    markEndStyle="url(#triangle)"
    onMouseLeave={[Function]}
    strokeOpacity={1}
  />
</g>
`;
